internal/server/handlers.go

```diff
@@ -115,6 +115,17 @@ type handler struct {
 	// If batchStarted is true, Notify will not notify notificationReceived. This allows the batch to
 	// complete before the response is sent, minimizing the number of responses.
 	batchStarted bool
+
+	// initialResourceVersions is a map of resource names to their initial version.
+	// this informs the server of the versions of the resources the xDS client knows of.
+	initialResourceVersions map[string]initialResourceVersion
```

Comment:
This needs to be a pointer, like this:
```suggestion
	initialResourceVersions map[string]*initialResourceVersion
```
Otherwise, this block doesn't do what you expect it to do:
```
	if res, ok := h.initialResourceVersions[name]; ok {
		if r != nil && res.version == r.Version {
			res.received = true
			return
		}
	}
```
Because `res` is not a reference to the value in the map, it's actually just a copy. It's the same reason why you can't do this:
```
m := map[int]int{"foo": 0}
i := m["foo"]
i++
```
`i` is a copy of the value in the map, not a reference, so `m` remains unchanged